haystack/nodes/label_generator/pseudo_label_generator.py

```diff
@@ -0,0 +1,216 @@
+from haystack.nodes.base import BaseComponent
+
+
+class PseudoLabelGenerator(BaseComponent):
+    """
+    The PseudoLabelGenerator is a component that creates Generative Pseudo Labeling (GPL) training data for training
+    of dense retrievers.
+
+    GPL is an unsupervised domain adaptation method for training of dense retrievers. It is based on question generation
+    and pseudo labeling with powerful cross-encoders. To train a domain-adapted model, it needs access to unlabeled
+    target corpus usually via DocumentStore and a retriever to mine for negatives.
+
+    For more details see https://github.com/UKPLab/gpl
+
+    For example:
+
+    ```python
+    |   document_store = DocumentStore(...)
+    |   retriever = Retriever(...)
+    |   qg = QuestionGenerator(model_name_or_path="doc2query/msmarco-t5-base-v1")
+    |   plg = PseudoLabelGenerator(qg, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    |
+    |   # Users can also provide a list of questions/document dictionary pairs to generate pseudo labels for
+    |
+    |   question_doc_pairs: List[Dict[str, str]] = [{"question": "question text ...", "document": "document text ..."}]
+    |   plg = PseudoLabelGenerator(question_doc_pairs, retriever)
+    |   train_examples = []
+    |   for idx, doc in enumerate(document_store):
+    |   output_samples = plg.run(documents=[doc])
+    |   for item in output_samples:
+    |       train_examples.append(item)
+    ```
+    """
+
+    def __init__(
+        self,
+        question_producer: Union[QuestionGenerator, List[Dict[str, str]]],
+        retriever: BaseRetriever,
+        cross_encoder_model_name_or_path: str = "cross-encoder/ms-marco-MiniLM-L-6-v2",
+        total_number_of_questions: int = sys.maxsize,
+        top_k: int = 10,
+    ):
+        """
+        Loads the cross encoder model and prepares PseudoLabelGenerator.
+
+        :param question_producer: The question producer used to generate questions or a list of already produced
+        questions/document pairs in Dict format {"question": "question text ...", "document": "document text ..."}.
+        :type question_producer: Union[QuestionGenerator, List[Dict[str, str]]]
+        :param retriever: The retriever used to query document stores
+        :type retriever: BaseRetriever
+        :param cross_encoder_model_name_or_path: The path to the cross encoder model, defaults to
+        cross-encoder/ms-marco-MiniLM-L-6-v2
+        :type cross_encoder_model_name_or_path: str (optional)
+        :param total_number_of_questions: This is the total number of questions that will be generated, defaults
+        to sys.maxsize
+        :type total_number_of_questions: int
+        :param top_k: This is the number of answers that you want to retrieve for each question, defaults to 10
```

Comment:
They use top_k = 50 in the paper, I feel like the default should also be 50 to stay close to the original implementation.

Reply:
hmm, 50 sounds a lot. I'll review it and lyk

Reply:
Confirmed. We'll use top_k = 50 as well. Thanks @mathislucka